File: Iteration2/SourceCode/src/main/java/Schedule.java
import java.util.ArrayList;

public class Schedule {
    private ArrayList<CourseSectionRegister> courseSectionRegisterList = new ArrayList<>();
    boolean isApproved;

    public Schedule(ArrayList<CourseSectionRegister> courseSectionRegisterList, boolean isApproved) {//+
        this.courseSectionRegisterList = courseSectionRegisterList;
        this.isApproved = isApproved;
    }

    public void addToSchedule(CourseSectionRegister courseSectionRegister){//
        ArrayList<CourseSectionRegister> courseSectionRegisterListToAdd=getCourseSectionRegisterList();
        courseSectionRegisterListToAdd.add(courseSectionRegister);
        
        setCourseSectionRegisterList(courseSectionRegisterListToAdd);
    }
    public void removeFromSchedule(CourseSectionRegister courseSectionRegister){//
        //ArrayList<CourseSectionRegister> courseSectionRegisterListToRemove=getCourseSectionRegisterList();
        //courseSectionRegisterListToRemove.remove(courseSectionRegister);
        //setCourseSectionRegisterList(courseSectionRegisterListToRemove);

        this.courseSectionRegisterList.remove(courseSectionRegister);
    }

    public boolean isApproved() {
        return isApproved;
    }//

    public void setApproved(boolean approved) {
        isApproved = approved;
    }//


    public ArrayList<CourseSectionRegister> checkCourseSectionCollision() {//
        ArrayList<CourseSectionRegister> ListOfCollision = new ArrayList<CourseSectionRegister>();
        CourseSectionRegister courseObject = null;

        ArrayList<String> FullDate = new ArrayList<String>();
        ArrayList<String> DayandFirstHourList = new ArrayList<String>();

        for (int i = 0; i < courseSectionRegisterList.size(); i++) {
            String[] splittedFromComma = courseSectionRegisterList.get(i).getCourseSection().getCourseHours().split(", ");
            for (String FullCourseDate : splittedFromComma) {
                FullDate.add(FullCourseDate);
            }
        }
        for (int i = 0; i < FullDate.size(); i++) {
            String[] DayandFirstHour = FullDate.get(i).split(":");
            DayandFirstHourList.add(DayandFirstHour[0]);
        }
        ArrayList<String> CheckList = CheckListIsDuplicated(DayandFirstHourList);

        for (int m = 0; m < CheckList.size(); m++) {
            for (int i = 0; i < courseSectionRegisterList.size(); i++) {
                if (courseSectionRegisterList.get(i).getCourseSection().getCourseHours().startsWith(CheckList.get(m))) {
                    if (courseObject == null) {
                        courseObject = courseSectionRegisterList.get(i);
                    } else if (courseSectionRegisterList.get(i).getCourseSection().getCourse().get(0).getSemester() > courseObject.getCourseSection().getCourse().get(0).getSemester()) {
                        ListOfCollision.add(courseSectionRegisterList.get(i));
                        courseObject = null;

                    } else {
                        ListOfCollision.add(courseObject);
                        courseObject = null;

                    }
                }
            }
        }

        return ListOfCollision;
    }
    public int calculateCreditTaken(){//
        int creditTaken =0;
        for(int i=0;i<courseSectionRegisterList.size();i++){
            creditTaken+=courseSectionRegisterList.get(i).getCourseSection().getCourse().get(0).getCredit();
        }
        return creditTaken;
    }

    public ArrayList<String> CheckListIsDuplicated(ArrayList<String> OnlyDay) {//
        ArrayList<String> CollisionedDaysList = new ArrayList<String>();

        for (int i = 0; i < OnlyDay.size(); i++) {
            for (int j = 0; j < OnlyDay.size(); j++) {
                if (i == j) continue; //same index position, ignore
                if (!CollisionedDaysList.contains(OnlyDay.get(i))) {
                    if (OnlyDay.get(i).equals(OnlyDay.get(j))) {
                        CollisionedDaysList.add(OnlyDay.get(i));
                    }
                }
            }
        }
        return CollisionedDaysList;
    }
    public ArrayList<CourseSectionRegister> getCourseSectionRegisterList() {
        return courseSectionRegisterList;
    }

    public void setCourseSectionRegisterList(ArrayList<CourseSectionRegister> courseSectionRegisterList) {//
        this.courseSectionRegisterList = courseSectionRegisterList;
    }
    public void approveSchedule(){//
        ArrayList<CourseSectionRegister> list=this.getCourseSectionRegisterList();
        for(CourseSectionRegister register:list){
            register.setStatus("Approved");
            register.getCourseSection().addToRegistrationList(register);
        }
        this.setApproved(true);
    }
    public void clearSchedule(){//
        getCourseSectionRegisterList().clear();
        this.setApproved(false);
    }
}

File: Iteration3/SourceCode/Schedule.py
class Schedule(object):

    def __init__(self, courseSectionRegisterList, isApproved):
        self.__courseSectionRegisterList = courseSectionRegisterList
        self.isApproved = isApproved

    def addToSchedule(self, courseSectionRegister):
        courseSectionRegisterListToAdd =self.getCourseSectionRegisterList()
        courseSectionRegisterListToAdd.append(courseSectionRegister)

        self.setCourseSectionRegisterList(courseSectionRegisterListToAdd)
    def removeFromSchedule(self, courseSectionRegister):
        if(courseSectionRegister in self.__courseSectionRegisterList):
            self.__courseSectionRegisterList.remove(courseSectionRegister)

    def isApproved(self):
        return self.isApproved

    def setApproved(self, approved):
        self.isApproved = approved


    def checkCourseSectionCollision(self):
        ListOfCollision = []
        courseObject = None
        FullDate = []
        DayandFirstHourList = []
        i = 0
        while i < len(self.__courseSectionRegisterList):
            splittedFromComma = self.__courseSectionRegisterList[i].getCourseSection().getCourseHours().split(", ")
            for FullCourseDate in splittedFromComma:
                FullDate.append(FullCourseDate)
            i += 1
        i = 0
        while i < len(FullDate):
            DayandFirstHour = FullDate[i].split(":")
            DayandFirstHourList.append(DayandFirstHour[0])
            i += 1
        CheckList = self.CheckListIsDuplicated(DayandFirstHourList)
        m = 0
        while m < len(CheckList):
            i = 0
            while i < len(self.__courseSectionRegisterList):
                if self.__courseSectionRegisterList[i].getCourseSection().getCourseHours().startswith(CheckList[m]):
                    if courseObject is None:
                        courseObject = self.__courseSectionRegisterList[i]
                    elif self.__courseSectionRegisterList[i].getCourseSection().getCourse()[0].getSemester() > courseObject.getCourseSection().getCourse()[0].getSemester():
                        ListOfCollision.append(self.__courseSectionRegisterList[i])
                        courseObject = None

                    else:
                        ListOfCollision.append(courseObject)
                        courseObject = None

                i += 1
            m += 1
        return ListOfCollision
    def calculateCreditTaken(self):
        creditTaken =0
        i = 0
        while i<len(self.__courseSectionRegisterList):
            creditTaken+=self.__courseSectionRegisterList[i].getCourseSection().getCourse()[0].getCredit()
            i += 1
        return creditTaken

    def CheckListIsDuplicated(self, OnlyDay):
        CollisionedDaysList = []
        i = 0
        while i < len(OnlyDay):
            j = 0
            while j < len(OnlyDay):
                if i == j:
                    j+=1
                    continue #same index position, ignore
                if not OnlyDay[i] in CollisionedDaysList:
                    if OnlyDay[i] == OnlyDay[j]:
                        CollisionedDaysList.append(OnlyDay[i])
                j += 1
            i += 1
        return CollisionedDaysList
    def getCourseSectionRegisterList(self):
        return self.__courseSectionRegisterList

    def setCourseSectionRegisterList(self, courseSectionRegisterList):
        self.__courseSectionRegisterList = courseSectionRegisterList
    def approveSchedule(self):
        list =self.getCourseSectionRegisterList()
        for register in list:
            register.setStatus("Approved")
            register.getCourseSection().addToRegistrationList(register)
        self.setApproved(True)
    def clearSchedule(self):
        self.getCourseSectionRegisterList().clear()
        self.setApproved(False)
